### ISSUE TITLE
target determination for functorch CI

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#82367]
### 🚀 The feature, motivation and pitch

I don't know if this would be worth it given that we're going to merge the functorch build system into PyTorch's build system in 1-2 months. But here goes:

Today, functorch depends on the pytorch package. pytorch does not depend on functorch. It would be cool if:
- if someone submitted a PR that only touched functorch, then we only run PyTorch builds, functorch builds, and functorch tests. I.e. we run no PyTorch tests.

This has the following benefits:
- it saves capacity
- it increases developer velocity for functorch-only devs

### Alternatives

functorch devs use `pytorchbot merge -f` with their best judgment

### Additional context

cc @seemethere @malfet @pytorch/pytorch-dev-infra

**Comment by janeyx99** [comment]
ccing @huydhn who is interested in TD

**Comment by huydhn** [comment]
Thank you for bringing this to my attention. This is one of the data point that I have been looking for actually :)

IMO, this is a simple way to reduce load and waiting time that we shouldn't ignore. I have seen several similar use cases in mind that don't really need to run the full testing workflow before merging, i.e. https://github.com/pytorch/pytorch/pull/82157. To be honest, they are all small and few and far between to get a high priority. 

Your use case, however, is much more important because it helps expedite the whole functorch development process. IMO, the proposal here would remain useful even after functorch moves into PyTorch's build system as long as the underlying assumption that pytorch does not depend on functorch stands (and it should be). So I think it's worth the effort to make this a higher priority.

Let me assign this to myself and do some further thinking on how to achieve this. As the next step, I'll look into functorch CI and reach out to the team to learn more (@zou3519, would you willing to be the POC?)

**Comment by janeyx99** [comment]
@malfet: would be good to work out a similar solution for mps tests!

**Comment by huydhn** [comment]
@zou3519 I have some updates on this that I want to get feedbacks from you. So I have reviewed some approaches with Dev Infra team today (Aug 11th).  Here is the details doc FYI https://docs.google.com/document/d/1IwwqpjXxDm-U5fMNz9e7ODzhfD-Z2vdxIv5kQ33qduo.

What do you think about the approach of having a functorch-related label attached to the PR to tell the CI to skip all tests besides functorch?  The workflow will be like this:

1. You guys submit PRs as usual
1. For functorch-only PRs, you can attach a label, for example `ciflow/functorch`.  Seeing that label, the CI will skip all the test shards besides functorch
1. For PRs that change both functorch and pytorch core, you don't need to do anything. The workflow runs all tests as usual
1. Once all the check finishes, tests that are not run will be shown as **SKIPPED**.  At the moment, we don't have a good way to filter out those skipped tests, but we will come back to this as the P1 goal

This approach is simple, so the MVP can be built quickly. If this works out, we could also think about building a new CI component on top of it to analyze the PR to automatically decide if it's functorch-only PR and attach the necessary labels.

What are your thoughts on this approach?

**Comment by zou3519** [comment]
This sounds reasonable to me. I think I prefer explicitly adding the label over something more automatic (as long as folks know what they're doing).

The assumption we're making here is "functorch depends on the pytorch package but not vice versa", but this might not be true for very long. When that time comes, would we just remove this feature, or would we try to figure out if it still makes sense to have target determination?

**Comment by zou3519** [comment]
Hmm, so a comment that I have (it may already be in the doc, I haven't gone through it) is: what happens when the PR is merged into trunk? Does it respect the 'ciflow: functorch' label, or would it run everything?

**Comment by huydhn** [comment]
> This sounds reasonable to me. I think I prefer explicitly adding the label over something more automatic (as long as folks know what they're doing).

Yeah, this is exactly the P0 for this change.  Devs will add the label explicitly.  We're happy with the initial assumption that folks know what they are doing.

> The assumption we're making here is "functorch depends on the pytorch package but not vice versa", but this might not be true for very long. When that time comes, would we just remove this feature, or would we try to figure out if it still makes sense to have target determination?

Yup, if that happens, functorch devs can just stop attaching the label to the PR.  We also plan to have a metric to record how people use this feature.  My guess is that when we see a drop in the functorch-related label, we will know that the above assumption is not valid anymore.

Actual automated target determination is a hard topic to get right.  It's in our [road map](https://docs.google.com/document/d/1IPLZAfgSUWe50qKkvV-Va0Hc7eaxikI3cyIN7dvO5IE), but work hasn't started yet.  So I think we will need to revisit this in the future. The good thing is the developer UX will most likely remain the same regardless of how this is achieved behind the scene.  We will definitely opt for a smoother developer experience as the ultimate goal.

**Comment by huydhn** [comment]
> Hmm, so a comment that I have (it may already be in the doc, I haven't gone through it) is: what happens when the PR is merged into trunk? Does it respect the 'ciflow: functorch' label, or would it run everything?

This is a very good observation. We discussed the 2 choices during the design review.  IMO, we should respect `ciflow/functorch` when merged into trunk and run only fucntorch tests. This is what I think we should aim for eventually because it's a logical thing to do for me (if we have unlimited capacity, we should run all the checks at PR, so why do pull and merge need to behave differently).  However, to achieve that, we will need to make some changes to the land validation workflow. So the consensus is to have that later after the label has been rolled out to pull request. 

TLDR;

1. Rollout manual test selection using labels to CI checks during pull request. Merge and land validation checks remain the same and run everything
2. Gather user feedback on the feature. See how confidence we are with it
3. Update merge and land validation check process to also respect the label

Does this sound like a good plan to you?

**Comment by huydhn** [comment]
@zou3519 I have the proof of concept change for this https://github.com/pytorch/pytorch/pull/83690 merged and ready.  To run only functorch tests, you will just need to attach [test-config/functorch](https://github.com/pytorch/pytorch/labels/test-config%2Ffunctorch) label into your PRs **BEFORE** the test starts.

It only supports Linux at the moment.  Windows and Mac support are on the way.  You can try it out right away without waiting for them.  Any comments are welcome :)

**Comment by zou3519** [comment]
> It only supports Linux at the moment. Windows and Mac support are on the way. You can try it out right away without waiting for them. Any comments are welcome :)

Cool, thank you, I will test!

**Comment by ZainRizvi** [comment]
Closing as done
